airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java

```diff
@@ -134,15 +134,33 @@ protected ChangeEvent<String, String> computeNext() {
 
       // if the last record matches the target file position, it is time to tell the producer to shutdown.
       if (targetPosition.reachedTargetPosition(eventAsJson)) {
-        LOGGER.info("Closing: Change event reached target position");
-        requestClose();
+        requestClose("Closing: Change event reached target position");
       }
       this.tsLastHeartbeat = null;
       this.lastHeartbeatPosition = null;
       this.receivedFirstRecord = true;
       this.maxInstanceOfNoRecordsFound = 0;
       return next;
     }
+
+    if (!signalledDebeziumEngineShutdown) {
+      LOGGER.warn("Debezium engine has not been signalled to shutdown, this is unexpected");
+    }
+
+    while (!debeziumShutdownProcedure.getRecordsRemainingAfterShutdown().isEmpty()) {
+      final ChangeEvent<String, String> event;
+      try {
+        event = debeziumShutdownProcedure.getRecordsRemainingAfterShutdown().poll(10, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+      if (event == null || isHeartbeatEvent(event)) {
+        continue;
+      }
+      hasSnapshotFinished = hasSnapshotFinished(Jsons.deserialize(event.value()));
+      return event;
+    }
+    throwExceptionIfSnapshotNotFinished();
```

Comment:
Why the process is not getting the events directly from `queue`?

Reply:
We need to make sure that the `queue` always has enough space but since we use a capacity queue, we need to read data from the queue in order to make space. The problem (as explained in the loom video attached in the description of the PR) is that when we trigger the shutdown of the Debezium engine, the main thread pauses until the engine shuts down. And since the reading from the `queue` is also happening in the main thread, the reading stops and if queue is already at capacity, it will not accept new records from Debezium. And Debezium wont shutdown gracefully cause it failed to hand over all the records that it fetched to the `queue`. 

As a solution now when we trigger shutdown, in order make sure the reading from the `queue` continues to happen, we start moving data from the `queue` in a separate new queue in a new thread. This allows debezium to shutdown gracefully, while the main thread continues to wait. Once the engine is shutdown, all we have to do is to read the remaining records that we moved to the new queue and we are good. 

**Please note that the main thread needs to be in pause state while debezium shuts down in order for the system to work without any edge cases.** And since the reading of the data happens inside the main thread, this the solution I chose. 

I did try another solution where we trigger the shutdown of the debezium engine in a separate thread and the main thread continues to run and read data but the implementation was even more complicated and the logic to track the state of the new thread was leaking all over the place (inside DebeziumRecordIterator and DebeziumRecordPublisher). Also it would have not worked cause of another new edge case that it would introduce. 

Right now the entire logic is contained inside the class `DebeziumShutdownProcedure` at one place which makes any changes to the other two classes in future easy to be done. 

I agree that this is not straight forward to understand but neither would any other solution. CDC itself is a complex piece of code. And anytime you are going to work with multi threading designs, things are complex.